### PR TITLE
Add a mechanism to register crash reporters

### PR DESCRIPTION
**Pull request by ChristophWurst**
This adds a crash reporter registry, which is meant for third party apps
to hook into the error logging/reporting chain. This way, external tools
like Sentry can be used to track and aggregate system crashes.

Out of personal interest (I'm using Sentry to monitor some Laravel apps), I looked into integrating Sentry into Nextcloud. As a first attempt, I just used the provided SDK and registered error/exception handlers. While this works for some triggered errors, it does not report and exception as we catch and handle them in places like `index.php`. Thus, I needed to find a way to hook into the exception reporting chain and found out that our logger is usually called to log the exception. Therefore, I've added a little mechanism that allows apps to register any reporter that implements a certain interface. If no reporters are registered, the registry doesn't do anything.

Usually, I would have opened a ticket first to discuss the general idea but since it's a rainy Sunday afternoon  and I was motivated to invest some time on this, I started the implementation right away and it was finished quickly.

Let me know what you think. The world doesn't end if anyone feels like this is too specific and we shouldn't add this to our public API. After all, someone has to maintain it. Still, I think it's a nice little enhancement and used correctly, the usage of Sentry will help detect more unnoticed errors that usually get lost in log files.

~~If we decide to integrate this I have to add some tests.~~ done

Steps to test
* Create an account on sentry.io or set up Sentry via Docker
* https://github.com/ChristophWurst/nextcloud_sentry
* Add the SDN via `php occ config:system:set sentry.dsn --value=<DSN>`
* Let something fail
* See the report on Sentry

cc @LukasReschke @nickvergessen @rullzer

### CODE REVIEW

File: lib/private/Log.php [hunk 1]
```diff
@@ -346,6 +351,7 @@ public function logException($exception, array $context = array()) {
 		$msg = isset($context['message']) ? $context['message'] : 'Exception';
 		$msg .= ': ' . json_encode($data);
 		$this->log($level, $msg, $context);
+		$this->crashReporters->delegateReport($exception);
```
Comment (ChristophWurst) [inline lib/private/Log.php]: - [ ] TODO: needs a null check

File: lib/private/Support/CrashReport/Registry.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	 * @since 13.0.0
+	 * @param IReporter $reporter
+	 */
+	public function register(IReporter $reporter) {
+		$this->reporters[] = $reporter;
+	}
+
+	/**
+	 * Delegate crash reporting to all registered reporters
+	 *
```
Comment (ChristophWurst) [inline lib/private/Support/CrashReport/Registry.php]: `@since` missing

File: lib/public/Support/CrashReport/IRegistry.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	 * Register a reporter instance
+	 *
+	 * @since 13.0.0
+	 * @param IReporter $reporter
+	 */
+	public function register(IReporter $reporter);
+
+	/**
+	 * Delegate crash reporting to all registered reporters
+	 *
```
Comment (ChristophWurst) [inline lib/public/Support/CrashReport/IRegistry.php]: `@since` missing

File: lib/public/Support/CrashReport/IReporter.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+use Throwable;
+
+/**
+ * @since 13.0.0
+ */
+interface IReporter {
+
+	/**
+	 * Report an (unhandled) exception
+	 *
```
Comment (ChristophWurst) [inline lib/public/Support/CrashReport/IReporter.php]: `@since` missing

**Comment by skjnldsv** [comment]
I don't really know specifically what are the full possibilities of sentry, but if I understand correctly, will this report automatically errors from users in the background?
Anonymously? :)

**Comment by ChristophWurst** [comment]
> I don't really know specifically what are the full possibilities of sentry, but if I understand correctly, will this report automatically errors from users in the background?

Exactly. Crashes (unhandled exceptions and errors) are sent to Sentry which collects and aggregates all reports, assigns them to users/releases and so on. This allows you to assign people to those errors and detect regressions.

bildschirmfoto von 2017-11-13 10-32-45


> Anonymously? :)

No, you have to either have an account on the hosted version or set it up yourself.

**Comment by skjnldsv** [comment]
Okay, so we can't just gather data from our users to report errors? (aside from the fact that those data could be sensitive and from the ethical view of this)

Looks promising either way! :)

**Comment by ChristophWurst** [comment]
> Okay, so we can't just gather data from our users to report errors? (aside from the fact that those data could be sensitive and from the ethical view of this)

Sure, we *could* do that. But I assume that takes some engineering power to do properly to actually gain anything from the anonymous reports ;)

**Comment by karlitschek** [comment]
great idea!  👍

**Comment by ChristophWurst** [comment]
I've added some tests. This should be ready to review :)

**Review by MorrisJobke — APPROVED** [review]
Tested, works and the code looks good 👍

**Comment by ChristophWurst** [comment]
> Tested, works and the code looks good 👍

\o/

I autosquashed the fixup commits and rebased onto master.

**Review by LukasReschke — APPROVED** [review]

**Pull request merged**